src/library/parsercsv.cpp

```diff
@@ -41,31 +41,33 @@ QList<QString> ParserCsv::parseAllLocations(const QString& playlistFile) {
 
         QList<QList<QString>> tokens = tokenize(bytes, ',');
 
-        // detect Location column
-        int locationColumnIndex = -1;
+        const auto detect_location_column =
+                [&](const auto& tokens_list,
+                        auto predicate) -> std::optional<std::size_t> {
+            const auto it = std::find_if(std::begin(tokens_list), std::end(tokens_list), predicate);
+            return (it != std::end(tokens_list))
+                    ? std::distance(std::begin(tokens_list), it)
+                    : std::optional<std::size_t>{};
+        };
         if (tokens.size()) {
-            for (int i = 0; i < tokens[0].size(); ++i) {
-                if (tokens[0][i] == QObject::tr("Location")) {
-                    locationColumnIndex = i;
-                    break;
-                }
-            }
-            if (locationColumnIndex < 0 && tokens.size() > 1) {
+            std::optional<std::size_t> locationColumnIndex = detect_location_column(
+                    tokens[0],
+                    [&](auto i) { return i == QObject::tr("Location"); });
+            if ((!locationColumnIndex.has_value()) && tokens.size() > 1) {
                 // Last resort, find column with path separators
                 // This happens in case of csv files in a different language
-                for (int i = 0; i < tokens[1].size(); ++i) {
-                    if (tokens[1][i].contains(QDir::separator())) {
-                        locationColumnIndex = i;
-                        break;
-                    }
-                }
+                locationColumnIndex = detect_location_column(tokens[1],
+                        [&](auto i) { return i.contains(QDir::separator()); });
             }
-            if (locationColumnIndex >= 0) {
-                for (int row = 1; row < tokens.size(); ++row) {
-                    if (locationColumnIndex < tokens[row].size()) {
-                        locations.append(tokens[row][locationColumnIndex]);
-                    }
-                }
+            if (locationColumnIndex.has_value()) {
+                std::for_each(std::next(std::begin(tokens)),
+                        std::end(tokens),
+                        [&](const auto& token) {
+                            if (locationColumnIndex < token.size()) {
+                                locations.append(token[static_cast<int>(
+                                        *locationColumnIndex)]);
+                            }
+                        });
```

Comment:
Mhmm. the code has just gotten more complex. I didn't anticipate that. My fault. Would you mind reverting the change again. My apology.

Reply:
Yeah :crying_cat_face: , the Mixxx project does not use the `range-v3` library yet (hint-hint! :wink:)... Then the code would be as simple as:
```cpp
for(const auto& token: ranges::views::tail(tokens)) {
  if (locationColumnIndex < token.size()) {
      locations.append(token[static_cast<int>(
              *locationColumnIndex)]);
  }
}
```
and many more nice things... :smiling_face_with_tear: _(`range-v3` stuff will arrive in `C++23`)_
I agree with you jointly that a range-based approach should be the default writing-style / default way to go.

Will revert back to the previous raw-for-loop form.

Reply:
Yeah... but expect to wait a bit for that... It was already a pain to switch to C++20...
You're welcome to add the `ranges-v3` library as a sort of polyfill to mixxx if you'd like. Do you have any good resources to recommend to learn about ranges? I don't know any mixxx maintainer that knows how to use them (me included).

Reply:
Interesting question @Swiftb0y, humm, it really depends on one's C++ knowledge.
The tl;dr form is:
1. Understanding Standard Library algorithms and why they are important (C++03 minimum).
2. Understanding the concept & reasoning behind [boost::range](https://www.boost.org/doc/libs/release/libs/range/) (Basically `range-v2`).
  (_Why is this `ranges-v2`? From the [History section of boost::range](https://www.boost.org/doc/libs/1_80_0/libs/range/doc/html/range/history_ack.html):_)
  `This version introduced Range Adaptors and Range Algorithms. This version 2 is the result of a merge of all of the RangeEx features into Boost.Range.`

  * 2.1 [boost::range - Introduction](https://www.boost.org/doc/libs/1_80_0/libs/range/doc/html/range/introduction.html)
  * 2.2 [boost::range - Range Adapter - Introduction & Motivation](https://www.boost.org/doc/libs/1_80_0/libs/range/doc/html/range/reference/adaptors/introduction.html)
  * 2.3 [boost::range - Range Algorithm - Introduction & Motivation](https://www.boost.org/doc/libs/1_80_0/libs/range/doc/html/range/reference/algorithms/introduction.html)
  * 2.4 Bonus Algorithms Resource: - [Christian Aichinger - Boost Range For Humans](https://www.caichinger.com/boost-range/index.html)
3. Understanding about [`range-v3`](https://github.com/ericniebler/range-v3)
  * [range-v3 - Quick Start](https://ericniebler.github.io/range-v3/index.html#tutorial-quick-start)
  * [Code Project - An Introduction to the Range-v3 Library](https://www.codeproject.com/Articles/5276756/An-Introduction-to-the-Range-v3-Library)

---
(Sadly,) This is a quick writeup, so thats all for now! Hope this helps and guides on your ranges-concept journey, ye? :smile_cat:

Reply:
Thanks a bunch. I'll definitely check those links out when I find the time.